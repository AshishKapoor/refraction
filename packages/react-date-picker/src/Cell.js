// @flow
import * as React from 'react';
import styled from '@emotion/styled/macro';
import css from '@emotion/css/macro';
import { textStyles, withFallback as wf } from '@quid/theme';
import { cx } from 'emotion';

import { CELL_HEIGHT, CELL_WIDTH } from './styles';

const Td = styled.td`
  height: ${CELL_HEIGHT + 6}px;
  width: ${CELL_WIDTH}px;
  vertical-align: middle;
  color: ${wf(props =>
    props.current ? props.theme.selected : props.theme.primary
  )};
  ${props =>
    textStyles(cx(props.current && 'bold', props.header && 'secondary'))};
`;

export const Button = styled.button`
  font: inherit;
  background: inherit;
  border: inherit;
  line-height: inherit;
  color: inherit;
  cursor: pointer;

  height: ${CELL_HEIGHT}px;
  width: ${CELL_WIDTH}px;
  border-radius: 2px;
  transition: all 0.2s ease-in-out;

  &:focus:not(:focus-visible) {
    outline: 0;
  }

  ${wf(
    props =>
      props.selected &&
      css`
        background-color: ${props.theme.current === 'light'
          ? props.theme.colors.gray4
          : props.theme.colors.white};
        color: ${props.theme.primaryInverse};
        ${textStyles('bold')(props)}
      `
  )}

  ${wf(
    props =>
      props.disabled &&
      css`
        cursor: default;
        color: ${props.theme.disabled};
      `
  )}
`;

type Props = {
  isCurrent?: boolean,
  isSelected?: boolean,
  isHeader?: boolean,
  onClick?: Function,
  children?: number | string,
  disabled?: boolean,
};

export default function Cell({
  isCurrent,
  isSelected,
  isHeader,
  onClick,
  children,
  disabled,
  ...props
}: Props) {
  return (
    <Td current={isCurrent} header={isHeader} {...props}>
      <Button
        type="button"
        onClick={onClick}
        disabled={disabled || isNaN(children)}
        selected={isSelected}
      >
        {children}
      </Button>
    </Td>
  );
}
