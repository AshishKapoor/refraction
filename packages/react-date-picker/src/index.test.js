// @flow
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { advanceTo } from 'jest-date-mock';
import Calendar from './';
import Cell from './Cell';

const WEEK_DAYS = 7;

const now = new Date(Date.UTC(2017, 0, 1));
advanceTo(now);

it('renders the expected markup', () => {
  const wrapper = shallow(
    <Calendar
      current={now}
      onChangeCurrent={f => f}
      onCancel={f => f}
      onSelect={f => f}
      selected={now}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it('works with no selection', () => {
  const wrapper = shallow(
    <Calendar
      current={now}
      onChangeCurrent={f => f}
      onCancel={f => f}
      onSelect={f => f}
    />
  );
  expect(
    wrapper.find('Row').findWhere(n => n.prop('selected') === undefined)
  ).toHaveLength(6);
});

it('navigates to next month', () => {
  const handleChangeCurrent = jest.fn();
  const wrapper = shallow(
    <Calendar
      current={now}
      onChangeCurrent={handleChangeCurrent}
      onCancel={f => f}
      onSelect={f => f}
      selected={now}
    />
  );
  wrapper
    .find('Button')
    .at(1)
    .simulate('click');
  expect(handleChangeCurrent.mock.calls[0][0].toDateString()).toBe(
    new Date(2017, 1, 1).toDateString()
  );
});

it('navigates to prev month', () => {
  const handleChangeCurrent = jest.fn();
  const wrapper = shallow(
    <Calendar
      current={now}
      onChangeCurrent={handleChangeCurrent}
      onCancel={f => f}
      onSelect={f => f}
      selected={now}
    />
  );
  wrapper
    .find('Button')
    .at(0)
    .simulate('click');
  expect(handleChangeCurrent.mock.calls[0][0].toDateString()).toBe(
    new Date(2016, 11, 1).toDateString()
  );
});

it('selects date on click on cell', () => {
  const handleSelect = jest.fn();
  const wrapper = mount(
    <Calendar
      current={now}
      onChangeCurrent={f => f}
      onCancel={f => f}
      onSelect={handleSelect}
      selected={now}
    />
  );
  wrapper
    .find(Cell)
    .at(WEEK_DAYS + 5)
    .find('button')
    .simulate('click');

  expect(handleSelect.mock.calls[0][0].toISOString().split('T')[0]).toBe(
    new Date(Date.UTC(2017, 0, 6)).toISOString().split('T')[0]
  );
});
