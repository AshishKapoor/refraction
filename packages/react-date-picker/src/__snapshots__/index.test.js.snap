// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the expected markup 1`] = `
<Container
  data-context="calendar"
  tabIndex={-1}
>
  <Header
    day={1}
    month={0}
    year={2017}
  />
  <Calendar>
    <Table>
      <thead>
        <tr>
          <Th
            colSpan={7}
          >
            <Navigator>
              <Button
                data-action="previous-month"
                importance="secondary"
                onClick={[Function]}
                size="small"
                transparent={true}
              >
                <Icon
                  name="chevron_left"
                  squared={true}
                />
              </Button>
              <Year>
                January
                 
                2017
              </Year>
              <Button
                data-action="next-month"
                importance="secondary"
                onClick={[Function]}
                size="small"
                transparent={true}
              >
                <Icon
                  name="chevron_right"
                  squared={true}
                />
              </Button>
            </Navigator>
          </Th>
        </tr>
        <tr>
          <Cell
            isHeader={true}
            key="0"
          >
            S
          </Cell>
          <Cell
            isHeader={true}
            key="1"
          >
            M
          </Cell>
          <Cell
            isHeader={true}
            key="2"
          >
            T
          </Cell>
          <Cell
            isHeader={true}
            key="3"
          >
            W
          </Cell>
          <Cell
            isHeader={true}
            key="4"
          >
            T
          </Cell>
          <Cell
            isHeader={true}
            key="5"
          >
            F
          </Cell>
          <Cell
            isHeader={true}
            key="6"
          >
            S
          </Cell>
        </tr>
      </thead>
      <Days>
        <Row
          date={2017-01-01T00:00:00.000Z}
          days={
            Array [
              1,
              2,
              3,
              4,
              5,
              6,
              7,
            ]
          }
          key="0"
          onSelect={[Function]}
          selected={2017-01-01T00:00:00.000Z}
        />
        <Row
          date={2017-01-01T00:00:00.000Z}
          days={
            Array [
              8,
              9,
              10,
              11,
              12,
              13,
              14,
            ]
          }
          key="1"
          onSelect={[Function]}
          selected={2017-01-01T00:00:00.000Z}
        />
        <Row
          date={2017-01-01T00:00:00.000Z}
          days={
            Array [
              15,
              16,
              17,
              18,
              19,
              20,
              21,
            ]
          }
          key="2"
          onSelect={[Function]}
          selected={2017-01-01T00:00:00.000Z}
        />
        <Row
          date={2017-01-01T00:00:00.000Z}
          days={
            Array [
              22,
              23,
              24,
              25,
              26,
              27,
              28,
            ]
          }
          key="3"
          onSelect={[Function]}
          selected={2017-01-01T00:00:00.000Z}
        />
        <Row
          date={2017-01-01T00:00:00.000Z}
          days={
            Array [
              29,
              30,
              31,
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          key="4"
          onSelect={[Function]}
          selected={2017-01-01T00:00:00.000Z}
        />
        <Row
          date={2017-01-01T00:00:00.000Z}
          days={
            Array [
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          key="5"
          onSelect={[Function]}
          selected={2017-01-01T00:00:00.000Z}
        />
      </Days>
    </Table>
  </Calendar>
</Container>
`;
